lib/charms/traefik_k8s/v0/ingress_per_unit.py

```diff
@@ -156,25 +249,177 @@ def __init__(self, charm: CharmBase, endpoint: str = None):
                 (defaults to "ingress-per-unit").
         """
         super().__init__(charm, endpoint)
-        self.framework.observe(self.on.ready, self._emit_request_event)
+        self.charm: CharmBase = charm
+        self.endpoint = endpoint
+
+        observe = self.framework.observe
+        rel_events = charm.on[endpoint]
+        observe(rel_events.relation_created, self._handle_relation)
+        observe(rel_events.relation_joined, self._handle_relation)
+        observe(rel_events.relation_changed, self._handle_relation)
+        observe(rel_events.relation_broken, self._handle_relation_broken)
+        observe(charm.on.leader_elected, self._handle_upgrade_or_leader)
+        observe(charm.on.upgrade_charm, self._handle_upgrade_or_leader)
+
+    @property
+    def app(self):
+        """Shortcut to self.charm.app."""
+        return self.charm.app
+
+    @property
+    def unit(self):
+        """Shortcut to self.charm.unit."""
+        return self.charm.unit
+
+    # @cached_property
+    @property
+    def relations(self):
+        """The list of Relation instances associated with this endpoint."""
+        return list(self.charm.model.relations[self.endpoint])
+
+    def _handle_relation(self, event):
+        if self.is_ready(event.relation):
+            self.on.ready.emit(event.relation)
+        elif self.is_available(event.relation):
+            self.on.available.emit(event.relation)
+        elif self.is_failed(event.relation):
+            self.on.failed.emit(event.relation)
+
+    @cache
```

Comment:
It makes it also hard(er) to test these methods, so I'm all +1 for removing it. I kept it because it was in SDI, and within that context it made sense because SDI managed its own status and could guarantee that it would never change within a charm's execution lifetime. But now that we own it, and if we think it might change in the future... Fair enough. Also it's not that big an optimization to warrant all the extra potential bugginess and mental overhead

Reply:
@mmanciop according to the same spirit, what'd you think about removing all caching in IPU and be done with it?

Reply:
+1